File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/SendOrderProjection.scala
package shopping.cart

import akka.actor.typed.ActorSystem
import akka.cluster.sharding.typed.ShardedDaemonProcessSettings
import akka.cluster.sharding.typed.scaladsl.ShardedDaemonProcess
import akka.persistence.query.Offset
import akka.persistence.r2dbc.query.scaladsl.R2dbcReadJournal
import akka.projection.eventsourced.scaladsl.EventSourcedProvider
import akka.projection.scaladsl.SourceProvider
import akka.projection.{ProjectionBehavior, ProjectionId}
import akka.persistence.query.typed.EventEnvelope
import akka.projection.Projection
import akka.projection.r2dbc.scaladsl.R2dbcProjection
import shopping.order.proto.ShoppingOrderService

object SendOrderProjection {

  def init(
            system: ActorSystem[_],
            orderService: ShoppingOrderService): Unit = {

    def sourceProvider(sliceRange: Range): SourceProvider[Offset, EventEnvelope[ShoppingCart.Event]] =
      EventSourcedProvider
        .eventsBySlices[ShoppingCart.Event](
          system,
          readJournalPluginId = R2dbcReadJournal.Identifier,
          "ShoppingCart",
          sliceRange.min,
          sliceRange.max)

    def projection(sliceRange: Range): Projection[EventEnvelope[ShoppingCart.Event]] = {
      val minSlice = sliceRange.min
      val maxSlice = sliceRange.max
      val projectionId = ProjectionId("SendOrderProjection", s"carts-$minSlice-$maxSlice")

      R2dbcProjection
        .atLeastOnce(
          projectionId,
          settings = None,
          sourceProvider(sliceRange),
          handler = () =>
            new SendOrderProjectionHandler(system, orderService))(system)
    }

    ShardedDaemonProcess(system).initWithContext(
      name = "SendOrderProjection",
      initialNumberOfInstances = 4,
      behaviorFactory = { daemonContext =>
        val sliceRanges =
          EventSourcedProvider.sliceRanges(system, R2dbcReadJournal.Identifier, daemonContext.totalProcesses)
        val sliceRange = sliceRanges(daemonContext.processNumber)
        ProjectionBehavior(projection(sliceRange))
      },
      ShardedDaemonProcessSettings(system),
      stopMessage = ProjectionBehavior.Stop)
  }

}

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-java/src/main/java/shopping/cart/SendOrderProjection.java
package shopping.cart;

import akka.actor.typed.ActorSystem;
import akka.cluster.sharding.typed.ShardedDaemonProcessSettings;
import akka.cluster.sharding.typed.javadsl.ShardedDaemonProcess;
import akka.japi.Pair;
import akka.persistence.query.Offset;
import akka.persistence.query.typed.EventEnvelope;
import akka.persistence.r2dbc.query.javadsl.R2dbcReadJournal;
import akka.projection.ProjectionBehavior;
import akka.projection.ProjectionId;
import akka.projection.eventsourced.javadsl.EventSourcedProvider;
import akka.projection.javadsl.ExactlyOnceProjection;
import akka.projection.javadsl.SourceProvider;

import java.util.List;
import java.util.Optional;

import akka.projection.r2dbc.R2dbcProjectionSettings;
import akka.projection.r2dbc.javadsl.R2dbcProjection;
import shopping.order.proto.ShoppingOrderService;

public class SendOrderProjection {

  private SendOrderProjection() {}

  public static void init(
      ActorSystem<?> system,
      ShoppingOrderService orderService) {
      ShardedDaemonProcess.get(system)
              .initWithContext( // <1>
                      ProjectionBehavior.Command.class,
                      "SendOrderProjection",
                      4,
                      daemonContext -> {
                          List<Pair<Integer, Integer>> sliceRanges =
                                  EventSourcedProvider.sliceRanges(system, R2dbcReadJournal.Identifier(), daemonContext.totalProcesses());
                          Pair<Integer, Integer> sliceRange =
                                  sliceRanges.get(daemonContext.processNumber());
                          return ProjectionBehavior.create(createProjection(system, orderService, sliceRange));
                      },
                      ShardedDaemonProcessSettings.create(system),
                      Optional.of(ProjectionBehavior.stopMessage()));
  }

    private static ExactlyOnceProjection<Offset, EventEnvelope<ShoppingCart.Event>>
    createProjection(
            ActorSystem<?> system,
            ShoppingOrderService shoppingOrderService,
            Pair<Integer, Integer> sliceRange) {

        int minSlice = sliceRange.first();
        int maxSlice = sliceRange.second();

        SourceProvider<Offset, EventEnvelope<ShoppingCart.Event>> sourceProvider =
                EventSourcedProvider.eventsBySlices(
                        system,
                        R2dbcReadJournal.Identifier(),
                        "ShoppingCart",
                        minSlice,
                        maxSlice);

        String slice = "carts-" + minSlice + "-" + maxSlice;
        Optional<R2dbcProjectionSettings> settings = Optional.empty();
        return R2dbcProjection.exactlyOnce( // <5>
                ProjectionId.of("SendOrderProjection", slice),
                settings,
                sourceProvider,
                () -> new SendOrderProjectionHandler(system, shoppingOrderService, slice),
                system);
    }
}
